File: 0202-happy-number/hashset_approach.py
# Time complexity: O(log N), Space complexity: O(log N)
class Solution:
    def isHappy(self, n: int) -> bool:
        visit = set()

        while n not in visit:
            visit.add(n)
            n = self.sumOfSquares(n)

            if n == 1:
                return True

        return False

    def sumOfSquares(self, n:int) -> int:
        tot = 0

        while n:
            dig = n%10
            dig = dig ** 2
            tot += dig
            n = n // 10

        return tot

File: 0202-happy-number/hashset_approach.cpp
// Time complexity: O(logn) as number of digits in a number = log n, Space complexity: O(log N)
class Solution {
public:
    bool isHappy(int n) {

        unordered_set<long long int> occurSums;
        occurSums.insert(n);
        long long int reqSum = n;
        
        while(reqSum != 1) {
            reqSum = compSqDigSum(reqSum);
            
            if (occurSums.count(reqSum)) {
                return false;
            }

            occurSums.insert(reqSum);
            
        }

        return true;
    }
private:
    long long int compSqDigSum(long long int num) {
        string strNum = to_string(num);
        long long int sqSum = 0;

        for (char ch: strNum) {
            int numInd = ch - '0';
            sqSum += numInd*numInd;
        }

        return sqSum;
    }
};